### PR TITLE
Issue #2020 disallow profile names with underscore

### LINKED ISSUE #2020: Unable to use underscore in profile name for centos iso
### General information

  * Minishift version: Version supporting profile 
  * OS: Linux / macOS / Windows
  * Hypervisor: KVM / xhyve / Hyper-V / VirtualBox
  * ISO : CentOS

### Steps to reproduce

  1. `minishift start --profile centos1_5 --iso-url centos`

### Expected

We should error out saying invalid profile name.

### Actual

Getting error.

### Logs

```
$ minishift start --profile centos1_5 --iso-url https://github.com/minishift/minishift-centos-iso/releases/download/v1.5.0/minishift-centos7.iso
-- Starting profile 'centos1_5'
-- Checking if requested hypervisor 'kvm' is supported on this platform ... OK
-- Checking if KVM driver is installed ... 
   Driver is available at /usr/local/bin/docker-machine-driver-kvm ... 
   Checking driver binary is executable ... OK
-- Checking if Libvirt is installed ... OK
-- Checking if Libvirt default network is present ... OK
-- Checking if Libvirt default network is active ... OK
-- Checking the ISO URL ... OK
-- Starting local OpenShift cluster using 'kvm' hypervisor ...
-- Minishift VM will be configured with ...
   Memory:    2 GB
   vCPUs :    2
   Disk size: 20 GB

   Downloading ISO 'https://github.com/minishift/minishift-centos-iso/releases/download/v1.5.0/minishift-centos7.iso'
 66.25 MiB / 324.00 MiB [==============================>-----------------------------------------------------------------------------------------------------------------------]  20.45% 8m12s 324.00 MiB / 324.00 MiB [========================================================================================================================================================] 100.00% 0s
-- Starting Minishift VM ................ FAIL E0213 13:22:20.689553   18400 start.go:368] Error starting the VM: Error creating the VM. Error creating machine: Error running provisioning: ssh command error:
command : sudo hostname centos1_5 && echo "centos1_5" | sudo tee /etc/hostname
err     : exit status 1
output  : hostname: the specified hostname is invalid
. Retrying.
Error starting the VM: Error creating the VM. Error creating machine: Error running provisioning: ssh command error:
command : sudo hostname centos1_5 && echo "centos1_5" | sudo tee /etc/hostname
err     : exit status 1
output  : hostname: the specified hostname is invalid
```

### PR DESCRIPTION
**Pull request by anjannath**
Fixes #2020 

Disallows the use of `_` in profile names, as centOS donot allow `_` in the hostname.

### CODE REVIEW

**Comment by budhrg** [comment]
@anjannath Need to rebase

**Comment by LalatenduMohanty** [comment]
retest this please

**Comment by anjannath** [comment]
I am not able to comprehend the integration test errors, will need @amitkrout 's help on this one. I am also unable to run the integration tests locally :(

**Comment by amitkrout** [comment]
@anjannath will check it and update ASAP

**Comment by amitkrout** [comment]
@anjannath For the time being can you please use correct [profile name](https://github.com/minishift/minishift/pull/2045/files#diff-dc1b91d3bab87ace73302cf020b00a11R59) ```foo``` instead of ```123foo```. This will resolve the failure. 

This failure is due to if same pattern is found in name for example ```foo``` and  ```123foo``` (here foo is the pattern). This needs factoring of the function [isTheActiveProfile](https://github.com/minishift/minishift/blob/master/test/integration/testsuite/minishift.go#L75).

**Comment by anjannath** [comment]
Thanks @amitkrout, All green now.

File: cmd/minishift/cmd/util/util.go [hunk 1]
```diff
@@ -109,6 +109,6 @@ func IsValidProfile(profileName string) bool {
 
 // IsValidProfileName return true if profile name follow ^[a-zA-Z0-9]+[\w+-]*$
 func IsValidProfileName(profileName string) bool {
-	rr := regexp.MustCompile(`^[a-zA-Z0-9]+[\w+-]*$`)
+	rr := regexp.MustCompile(`^[a-zA-Z0-9]+[a-zA-Z0-9-]*$`)
```
Comment (praveenkumar) [inline cmd/minishift/cmd/util/util.go]: :+1:

File: cmd/minishift/cmd/util/util_test.go [hunk 2]
```diff
@@ -29,14 +29,14 @@ func TestIsValidProfileName(t *testing.T) {
 	}{{"test", true},
 		{".", false},
 		{"@foo", false},
+		{"_mini", false},
 		{"foo123", true},
 		{"123foo", true},
-		{"foo_123", true},
+		{"foo_123", false},
```
Comment (praveenkumar) [inline cmd/minishift/cmd/util/util_test.go]: :+1:

File: docs/source/using/profiles.adoc [hunk 3]
```diff
@@ -39,7 +39,7 @@ There are two ways to create a new profile.
 [NOTE]
 ====
 Profile name can only consist of alphanumeric characters.
-The use of underscore ( *_* ) and hyphen ( *-* ) are allowed as separator.
+The use of hyphen ( *-* ) is allowed as separator. The regex `^[a-zA-Z0-9]+[a-zA-Z0-9-]*$` is used to validate profile names.
```
Comment (robin-owen) [inline docs/source/using/profiles.adoc]: I'm not sure that we need to include the regular expression used within the documentation. We're already describing which characters are valid for profile names, and I think that's ideal. :)

Basically, I would remove the added sentence -- I'm not convinced that it really contributes anything to the documentation.

Comment (amitkrout) [reply docs/source/using/profiles.adoc]: > Basically, I would remove the added sentence -- I'm not convinced that it really contributes anything to the documentation.

+1. Its really not good to add the code behind in doc. This sentence ```The regex `^[a-zA-Z0-9]+[a-zA-Z0-9-]*$` is used to validate profile names.``` should be completely removed.

File: cmd/minishift/cmd/util/util_test.go (hunk 2 above)
Comment (amitkrout) [reply cmd/minishift/cmd/util/util_test.go]: +1

File: docs/source/using/profiles.adoc (hunk 3 above)
Comment (praveenkumar) [reply docs/source/using/profiles.adoc]: -1 for ` The regex `^[a-zA-Z0-9]+[a-zA-Z0-9-]*$` is used to validate profile names.` in the doc section. We already say `Profile name can only consist of alphanumeric characters.`

Comment (anjannath) [reply docs/source/using/profiles.adoc]: Alright.. removing it!

File: docs/source/using/profiles.adoc [hunk 4]
```diff
@@ -39,7 +39,7 @@ There are two ways to create a new profile.
 [NOTE]
 ====
 Profile name can only consist of alphanumeric characters.
-The use of underscore ( *_* ) and hyphen ( *-* ) are allowed as separator.
+The use of hyphen ( *-* ) is allowed as separator.
```
Comment (praveenkumar) [inline docs/source/using/profiles.adoc]: :+1:

**Comment by LalatenduMohanty** [comment]
Thanks @anjannath .

**Pull request merged**